drivers/bluetooth/hci/rpmsg_nrf53.c

```diff
@@ -0,0 +1,241 @@
+{
+	BT_DBG("Received message of %u bytes.", len);
+	BT_HEXDUMP_DBG((uint8_t *)data, len, "Data:");
+
+	bt_rpmsg_rx(data, len);
+
+	return RPMSG_SUCCESS;
+}
+
+static void rpmsg_service_unbind(struct rpmsg_endpoint *ep)
+{
+	rpmsg_destroy_ept(ep);
+}
+
+void ns_bind_cb(struct rpmsg_device *rdev, const char *name, u32_t dest)
+{
+	(void)rpmsg_create_ept(&ep,
+				rdev,
+				name,
+				RPMSG_ADDR_ANY,
+				dest,
+				endpoint_cb,
+				rpmsg_service_unbind);
+
+	k_sem_give(&sync_sem);
+}
+
+int bt_rpmsg_platform_init(void)
+{
+	int err;
+	struct metal_init_params metal_params = METAL_INIT_DEFAULTS;
+
+	static struct virtio_vring_info     rvrings[2];
+	static struct rpmsg_virtio_shm_pool shpool;
+	static struct virtio_device         vdev;
+	static struct rpmsg_virtio_device   rvdev;
+	static struct metal_io_region       *io;
+	static struct metal_device          *device;
+
+	/* Libmetal setup */
+	err = metal_init(&metal_params);
+	if (err) {
+		BT_ERR("metal_init: failed - error code %d", err);
+		return err;
+	}
+
+	err = metal_register_generic_device(&shm_device);
+	if (err) {
+		BT_ERR("Couldn't register shared memory device: %d", err);
+		return err;
+	}
+
+	err = metal_device_open("generic", SHM_DEVICE_NAME, &device);
+	if (err) {
+		BT_ERR("metal_device_open failed: %d", err);
+		return err;
+	}
+
+	io = metal_device_io_region(device, 0);
+	if (!io) {
+		BT_ERR("metal_device_io_region failed to get region");
+		return -ENODEV;
+	}
+
+	/* IPM setup */
+	ipm_tx_handle = device_get_binding("IPM_0");
+	if (!ipm_tx_handle) {
+		BT_ERR("Could not get TX IPM device handle");
+		return -ENODEV;
+	}
+
+	ipm_rx_handle = device_get_binding("IPM_1");
+	if (!ipm_rx_handle) {
+		BT_ERR("Could not get RX IPM device handle");
+		return -ENODEV;
+	}
+
+	ipm_register_callback(ipm_rx_handle, ipm_callback, NULL);
+
+	/* Virtqueue setup */
+	vq[0] = virtqueue_allocate(VRING_SIZE);
+	if (!vq[0]) {
+		BT_ERR("virtqueue_allocate failed to alloc vq[0]");
+		return -ENOMEM;
+	}
+
+	vq[1] = virtqueue_allocate(VRING_SIZE);
+	if (!vq[1]) {
+		BT_ERR("virtqueue_allocate failed to alloc vq[1]");
+		return -ENOMEM;
+	}
+
+	rvrings[0].io = io;
+	rvrings[0].info.vaddr = (void *)VRING_TX_ADDRESS;
+	rvrings[0].info.num_descs = VRING_SIZE;
+	rvrings[0].info.align = VRING_ALIGNMENT;
+	rvrings[0].vq = vq[0];
+
+	rvrings[1].io = io;
+	rvrings[1].info.vaddr = (void *)VRING_RX_ADDRESS;
+	rvrings[1].info.num_descs = VRING_SIZE;
+	rvrings[1].info.align = VRING_ALIGNMENT;
+	rvrings[1].vq = vq[1];
+
+	vdev.role = RPMSG_MASTER;
+	vdev.vrings_num = VRING_COUNT;
+	vdev.func = &dispatch;
+	vdev.vrings_info = &rvrings[0];
+
+	rpmsg_virtio_init_shm_pool(&shpool, (void *)SHM_START_ADDR, SHM_SIZE);
+	err = rpmsg_init_vdev(&rvdev, &vdev, ns_bind_cb, io, &shpool);
+	if (err) {
+		BT_ERR("rpmsg_init_vdev failed %d", err);
+		return err;
+	}
+
+	/* Since we are using name service, we need to wait for a response
+	 * from NS setup and than we need to process it
+	 */
+	virtqueue_notification(vq[0]);
```

Comment:
This seems useless. when called remote processor still not ready to send the NS message, so nothing to process in queue.
Application will be blocked on sync_sem waiting rpmsg from the remote side. 
ipm_callback should call the virtqueue_notification that will process message and call the ns_bind_cb

Reply:
You are right - this is redundant.

Reply:
fixed